In [196]:
from pandas import DataFrame
import pandas as pd
import numpy as np

import pyodbc
from sqlalchemy import create_engine

db_file = r'C:\Users\user\Downloads\2.7 Исходные данные\Копия 2021.accdb'
user = 'user'
password = 'pw'

odbc_conn_str = 'DRIVER={Microsoft Access Driver (*.mdb, *.accdb)};DBQ=%s;UID=%s;PWD=%s' % (db_file, user, password)
conn = pyodbc.connect(odbc_conn_str)

cur = conn.cursor()


query = "SELECT * FROM 2021"
dataf2021 = pd.read_sql(query, conn)
conn.close()
dataf2021[['Year', 'Number of stores','Month']] = dataf2021[['Year', 'Number of stores','Month']].astype(int)
dataf2021['Brand'].mask(dataf2021['Brand'] == 'Greennn#!', 'Green', inplace=True)
dataf2021[['Year', 'Number of stores','Month']] = dataf2021[['Year', 'Number of stores','Month']].astype(int)
dataf2021['Value (in 1000 rub)'] = dataf2021['Value (in rub)'].apply(lambda x: x/1000)
dataf2021 = dataf2021.drop('Value (in rub)', axis=1)
dataf2021['Value (in 1000 rub)'] = dataf2021['Value (in 1000 rub)'].astype(float)

C:\Users\user\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [197]:
dataf2021

Year  Month       Channel  Brand Weight range  Units (in 1000)  \
0    2021      1  Hypermarkets  Green     200-249G          12.7583   
1    2021      1  Hypermarkets  Green     350-399G          61.6034   
2    2021      1  Hypermarkets  Green     400-599G          47.7524   
3    2021      1  Hypermarkets  Green     600-899G           0.0017   
4    2021      1  Hypermarkets   Blue     200-249G           0.0509   
..    ...    ...           ...    ...          ...              ...   
670  2021     11   Minimarkets  Green     100-199G          11.4728   
671  2021     11  Supermarkets  Green     100-199G          17.9486   
672  2021     12  Hypermarkets  Green     100-199G          10.5903   
673  2021     12   Minimarkets  Green     100-199G           7.8182   
674  2021     12  Supermarkets  Green     100-199G          25.2956   

     Volume (in 1000 kg)  Number of stores  Value (in 1000 rub)  
0                 2.6792               545           817.789128  
1                21.5612               798          5598.975814  
2                23.8763               569          4638.537271  
3                 0.0013                 1             0.123992  
4                 0.0101                 6             5.628921  
..                   ...               ...                  ...  
670               2.2945              1710           749.057600  
671               3.5896              2387          1339.453100  
672               2.1180               428           748.857900  
673               1.5636              1053           549.344300  
674               5.0591              2407          1721.988600  

[675 rows x 9 columns]

In [199]:
dataf2020 = pd.read_csv('Копия 2020.txt')
dataf2020['Number of stores'] = dataf2020['Number of stores'].fillna(0)
dataf2020['Month'] = dataf2020['Month'].fillna(0)
dataf2020[['Year', 'Number of stores','Month']] = dataf2020[['Year', 'Number of stores','Month']].astype(int)
dataf2020_1 = dataf2020[dataf2020['Volume (in 1000 kg)'] != 1000.0000]
dataf2020_1['Value (in 1000 rub)'] = dataf2020_1['Value (in 1000 rub)'].astype(float)

C:\Users\user\AppData\Local\Temp/ipykernel_10180/4222784131.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataf2020_1['Value (in 1000 rub)'] = dataf2020_1['Value (in 1000 rub)'].astype(float)


In [200]:
dataf2020_1

Year  Month       Channel  Brand Weight range  Units (in 1000)  \
0    2020      1  Hypermarkets  Green     200-249G          16.3202   
1    2020      1  Hypermarkets  Green     350-399G          87.8634   
2    2020      1  Hypermarkets  Green     400-599G          35.7182   
3    2020      1  Hypermarkets  Green     600-899G           0.3580   
4    2020      1  Hypermarkets   Blue     400-599G          16.8368   
..    ...    ...           ...    ...          ...              ...   
663  2020     11   Minimarkets  Green     100-199G          17.7682   
664  2020     11  Supermarkets  Green     100-199G          18.8734   
665  2020     12  Hypermarkets  Green     100-199G          36.9079   
666  2020     12   Minimarkets  Green     100-199G          20.2285   
667  2020     12  Supermarkets  Green     100-199G          40.6617   

     Value (in 1000 rub)  Volume (in 1000 kg)  Number of stores  
0             936.803413               3.4272               477  
1            7019.116508              30.7522               754  
2            4166.453788              17.8592               629  
3              66.387690               0.2685                29  
4            3227.992656               6.7346               663  
..                   ...                  ...               ...  
663          1109.393100               3.5537              1713  
664          1317.491600               3.7747              1607  
665          2267.318300               7.3816               468  
666          1371.255000               4.0456              1889  
667          2642.164800               8.1324              2079  

[668 rows x 9 columns]

In [207]:
dataf2022 = pd.read_csv('Копия 2022.txt', delimiter = '\t')
dataf2022[['Year', 'Number of stores','Month']] = dataf2022_1[['Year', 'Number of stores','Month']].astype(int)
dataf2022['Value (in 1000 rub)'] = dataf2022['Value (in 1000 rub)'].str.replace("[+,]", "", regex=True)
dataf2022['Value (in 1000 rub)'] = dataf2022['Value (in 1000 rub)'].astype(float)
dataf2022['Value (in 1000 rub)'] = dataf2022['Value (in 1000 rub)'].fillna(0)

In [208]:
dataf2022

Year  Month       Channel  Brand Weight range  Units (in 1000)  \
0    2022      1  Hypermarkets  Green     200-249G          15.2249   
1    2022      1  Hypermarkets  Green     350-399G          58.2443   
2    2022      1  Hypermarkets  Green     400-599G          38.4102   
3    2022      1  Hypermarkets   Blue     200-249G           0.1327   
4    2022      1  Hypermarkets   Blue     400-599G          44.9287   
..    ...    ...           ...    ...          ...              ...   
676  2022     11   Minimarkets  Green     100-199G           1.1740   
677  2022     11  Supermarkets  Green     100-199G           4.2365   
678  2022     12  Hypermarkets  Green     100-199G           0.1749   
679  2022     12   Minimarkets  Green     100-199G           2.0995   
680  2022     12  Supermarkets  Green     100-199G           2.5092   

     Value (in 1000 rub)  Volume (in 1000 kg)  Number of stores  
0             899.265493               3.1972               536  
1            5622.580442              20.3855               758  
2            4457.224790              19.2051               633  
3              15.948333               0.0265                10  
4            7611.563208              17.9715               728  
..                   ...                  ...               ...  
676            97.225600               0.2348               187  
677           353.291000               0.8473               778  
678            14.340600               0.0350                22  
679           123.600800               0.4200               200  
680           223.190100               0.5018               514  

[681 rows x 9 columns]

In [209]:
data_frames = [dataf2020_1, dataf2021, dataf2022]

In [210]:
result = pd.concat(data_frames)

### 3.7 Практическая работа

Цели практической работы
Научиться рассчитывать бизнес-метрики на основе обработанных данных.
Понять, чем отличается каждая из динамических метрик и какую информацию она даёт.
Научиться определять структуру рынка.
Проиллюстрировать различные срезы данных, через которые метрики могут быть проанализированы.
Развить способность делать выводы на основе полученных результатов.

Что нужно сделать
Задание 1. Динамика роста продаж в деньгах (данные в столбце 'Value (in 1000 rub)') бренда Green (фильтр по столбцу 'Brand') за весь 2022 год относительно 2021 года (значения в столбце 'Year' = 2022 относительно второго в 2021), 

Задание 2. рост продаж за последние три месяца 

Задание 3. и за последний месяц.

Задание 4. Также необходимо сравнить эти показатели с аналогичными цифрами всего рынка. (фильтр по столбцу 'Brand' ВСЕ, за исключением Green)

Задание 5. Исторически бренд Green занимал лидирующую позицию на рынке, поэтому важно построить рейтинг топ-5 брендов категории на основе данных продаж в рублях за все три года.

In [224]:
result[result['Brand'] == 'Green'].groupby(['Brand','Year'])['Value (in 1000 rub)'].sum()

Brand  Year
Green  2020    496393.397464
       2021    522684.061065
       2022    500728.413849
Name: Value (in 1000 rub), dtype: float64

In [231]:
Green_in_2022 = result[result['Brand'] == 'Green'].groupby(['Brand','Year'])['Value (in 1000 rub)'].sum()[2]
Green_in_2021 = result[result['Brand'] == 'Green'].groupby(['Brand','Year'])['Value (in 1000 rub)'].sum()[1]
print(f'Ответ на Задание 1 : Динамика продаж в деньгах составила {(1 - Green_in_2022/Green_in_2021) * 100 :.1f}%')

Ответ на Задание 1 : Динамика продаж в деньгах составила 4.2%


In [233]:
result[(result['Brand'] == 'Green') & (result['Year'] == 2022)].groupby(['Brand','Month'])['Value (in 1000 rub)'].sum()

Brand  Month
Green  1        50535.553419
       2        47166.585859
       3        58961.403624
       4        46472.727198
       5        42935.991587
       6        35427.072950
       7        37050.243842
       8        35029.654804
       9        32881.882497
       10       44702.156656
       11       34353.067849
       12       35212.073564
Name: Value (in 1000 rub), dtype: float64

In [246]:
Green_in__22_3m = result[(result['Brand'] == 'Green') & (result['Year'] == 2022) & ((result['Month'] >= 10))]['Value (in 1000 rub)'].sum()
Green_in__22_before_3m = result[(result['Brand'] == 'Green') & (result['Year'] == 2022) & (result['Month'] <= 9) & (result['Month'] >= 7)]['Value (in 1000 rub)'].sum()

print(f'Ответ на Задание 2. Изменение продаж за последние три месяца {(1 - Green_in__22_3m/Green_in__22_before_3m) * 100 :.1f}%')


Ответ на Задание 2. Изменение продаж за последние три месяца -8.9%


In [247]:
Green_in__22_1m = result[(result['Brand'] == 'Green') & (result['Year'] == 2022) & ((result['Month'] >= 12))]['Value (in 1000 rub)'].sum()
Green_in__22_before_1m = result[(result['Brand'] == 'Green') & (result['Year'] == 2022) & (result['Month'] <= 11) & (result['Month'] >= 11)]['Value (in 1000 rub)'].sum()

print(f'Ответ на Задание 3. Изменение продаж за последний месяц {(1 - Green_in__22_1m/Green_in__22_before_1m) * 100 :.1f}%')

Ответ на Задание 3. Изменение продаж за последний месяц -2.5%


In [284]:
Brands = ['Green', 'Blue', 'Orange','Yellow','Black','Deep', 'Red', 'White', 'Rosy', 'Dim', 'Pale', 'Vivid', 'Magenta']
Green_in_2022 = result[(result['Brand'] == 'Green') & (result['Year'] == 2022)]['Value (in 1000 rub)'].sum()
Green_in_2021 = result[(result['Brand'] == 'Green') & (result['Year'] == 2021)]['Value (in 1000 rub)'].sum()
dinamic_green = (1 - Green_in_2022/Green_in_2021) * 100
for _ in Brands :
    Green_in_2022 = result[(result['Brand'] == _) & (result['Year'] == 2022)]['Value (in 1000 rub)'].sum()
    Green_in_2021 = result[(result['Brand'] == _) & (result['Year'] == 2021)]['Value (in 1000 rub)'].sum()
    dinamic = (1 - Green_in_2022/Green_in_2021) * 100
    if (dinamic > dinamic_green) :
        print(f'Динамика продаж в деньгах для {_} составила {dinamic :.1f}% и лучше Green')

Динамика продаж в деньгах для Black составила 95.4% и лучше Green
Динамика продаж в деньгах для Red составила 28.4% и лучше Green
Динамика продаж в деньгах для Rosy составила 64.9% и лучше Green
Динамика продаж в деньгах для Dim составила 40.7% и лучше Green
Динамика продаж в деньгах для Pale составила 100.0% и лучше Green
Динамика продаж в деньгах для Magenta составила 62.4% и лучше Green


In [287]:
print(f'Для топ 10 брендов у 6 из них динамика роста продаж лучше чем у Green {dinamic_green :.1f} %.')

Для топ 10 брендов у 6 из них динамика роста продаж лучше чем у Green 4.2 %.


In [293]:
Brands = ['Green', 'Blue', 'Orange','Yellow','Black','Deep', 'Red', 'White', 'Rosy', 'Dim', 'Pale', 'Vivid', 'Magenta']
print('Динамика 2021 года по сравнению с 2020')
brands_top = []
for _ in Brands :
    Green_in_2021 = result[(result['Brand'] == _) & (result['Year'] == 2021)]['Value (in 1000 rub)'].sum()
    Green_in_2020 = result[(result['Brand'] == _) & (result['Year'] == 2020)]['Value (in 1000 rub)'].sum()
    if Green_in_2020 != 0:
        dinamic = (1 - Green_in_2021/Green_in_2020) * 100
        print(f'Динамика продаж в деньгах для {_} составила {dinamic :.1f}%')
        brands_top.append(dinamic)

Динамика 2021 года по сравнению с 2020
Динамика продаж в деньгах для Green составила -5.3%
Динамика продаж в деньгах для Blue составила -18.9%
Динамика продаж в деньгах для Orange составила -38.4%
Динамика продаж в деньгах для Yellow составила 14.2%
Динамика продаж в деньгах для Black составила 65.9%
Динамика продаж в деньгах для Deep составила 7.5%
Динамика продаж в деньгах для Red составила 57.1%
Динамика продаж в деньгах для White составила -142.6%
Динамика продаж в деньгах для Dim составила 4.7%
Динамика продаж в деньгах для Pale составила 88.3%
Динамика продаж в деньгах для Vivid составила 51.9%


In [295]:
print('ТОП5 брендов по росту продаж: Yellow, Black, Red, Pale, Vivid')

ТОП5 брендов по росту продаж: Yellow, Black, Red, Pale, Vivid


In [281]:
print('Динамика 2022 года по сравнению с 2021')
for _ in Brands :
    Green_in_2021 = result[(result['Brand'] == _) & (result['Year'] == 2022)]['Value (in 1000 rub)'].sum()
    Green_in_2020 = result[(result['Brand'] == _) & (result['Year'] == 2021)]['Value (in 1000 rub)'].sum()
    if Green_in_2020 != 0:
        dinamic = (1 - Green_in_2021/Green_in_2020) * 100
        print(f'Динамика продаж в деньгах для {_} составила {dinamic :.1f}%')

Динамика 2022 года по сравнению с 2021
Динамика продаж в деньгах для Green составила 4.2%
Динамика продаж в деньгах для Blue составила -35.4%
Динамика продаж в деньгах для Orange составила -495.1%
Динамика продаж в деньгах для Yellow составила 2.6%
Динамика продаж в деньгах для Black составила 95.4%
Динамика продаж в деньгах для Deep составила -12.2%
Динамика продаж в деньгах для Red составила 28.4%
Динамика продаж в деньгах для White составила -11.7%
Динамика продаж в деньгах для Rosy составила 64.9%
Динамика продаж в деньгах для Dim составила 40.7%
Динамика продаж в деньгах для Pale составила 100.0%
Динамика продаж в деньгах для Vivid составила -22.8%
Динамика продаж в деньгах для Magenta составила 62.4%


In [296]:
print('ТОП5 брендов по росту продаж: Black, Rosy, Dim, Pale, Magenta')

ТОП5 брендов по росту продаж: Black, Rosy, Dim, Pale, Magenta


# Выводы:

Ответ на Задание 1 : Динамика продаж в деньгах составила 4.2%
Ответ на Задание 2. Изменение продаж за последние три месяца -8.9%
Ответ на Задание 3. Изменение продаж за последний месяц -2.5%
Для топ 10 брендов у 6 из них динамика роста продаж лучше чем у Green 4.2 %.
ТОП5 брендов по росту продаж 2021: Yellow, Black, Red, Pale, Vivid
ТОП5 брендов по росту продаж 2022: Black, Rosy, Dim, Pale, Magenta